idaes_examples/notebooks/docs/scaling/scaler_workshop.ipynb

```diff
@@ -0,0 +1,1944 @@
+   "id": "85175874",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "====================================================================================\n",
+      "The following variable(s) are associated with extreme Jacobian values (<1.0E-04 or>1.0E+04):\n",
+      "\n",
+      "    fs.equil.control_volume.properties_out[0.0].flow_vol: 9.427E+07\n",
+      "    fs.equil.control_volume.properties_out[0.0].temperature: 4.172E+06\n",
+      "    fs.equil.control_volume.rate_reaction_extent[0.0,R1]: 4.900E+04\n",
+      "\n",
+      "====================================================================================\n",
+      "====================================================================================\n",
+      "The following constraint(s) are associated with extreme Jacobian values (<1.0E-04 or>1.0E+04):\n",
+      "\n",
+      "    fs.equil.control_volume.enthalpy_balances[0.0]: 9.436E+07\n",
+      "    fs.equil.control_volume.material_balances[0.0,Liq,H2O]: 5.539E+04\n",
+      "\n",
+      "====================================================================================\n"
+     ]
+    }
+   ],
+   "source": [
+    "dt.display_variables_with_extreme_jacobians()\n",
+    "dt.display_constraints_with_extreme_jacobians()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "777b15b6",
+   "metadata": {},
+   "source": [
+    "These diagnostics can help give us an idea of what may be causing problems in our model. From the output above, we can see that the variables with large Jacobian norms (i.e., high sensitivities) are the outlet flow rate and temperature, as well as the rate-based extent of reaction. We can also see that the constraints with large Jacobian norms are the enthalpy balance and H20 material balance for the reactor. Understanding what this means however is often complicated and requires some thought.\n",
+    "\n",
+    "For example, one might wonder why the volumetric flow rate at the outlet of the reactor is so important as it is effectively determined by the inlet flow rate (due to the water balance effectively conserving volume). However, it is important to realize that from the point of view of the solver, the outlet flowrate is a free variable that can be manipulated, and that changing the volumetric flowrate at the outlet has a significant effect on the outlet concentrations (all other things kept constant). Thus as far as the solver is concerned, the outlet volumetric flowrate has a significant impact on the solution. This also explains why the H2O material balance is so important, as this is what governs the outlet flowrate. However, the material balances and concentrations of the other species in the system are what really matter, but this is not reflected in the model Jacobian. Once of the goals of scaling the model should be to balance these to ensure that the water balance does not come to dominate the other species.\n",
```

Comment:
If you want to understand where those values are coming from, you need to use the method in the `SVDToolbox` to print out Jacobian rows and columns. You'll quickly find that, for the volumetric flowrate, it has an enormous Jacobian value because it's being multiplied by the molar enthalpy (1e3 to 1e4 in J/mol) and the molar density (5.5e4 mol/m^3 for water) in the enthalpy balance constraint.

The volumetric flow rate being determined by the inlet flow rate is immaterial. Even if it is, changes in the volumetric flowrate to satisfy that constraint must be counteracted by other variables to make the residual small. And the volumetric flowrate _does_ have a significant impact on the solution. The impact is just seen in variables like enthalpy, not the ones users typically look at.

I think it's worth mentioning that having a big value in the Jacobian does not mean a variable is "important". It's how sensitive the constraint residual is to that change in variable.

Reply:
Fixed.